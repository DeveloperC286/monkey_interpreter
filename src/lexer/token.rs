#[derive(PartialEq, PartialOrd)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

#[derive(PartialEq, PartialOrd)]
#[allow(non_camel_case_types)]
pub enum TokenType {
    ILLEGAL,
    EOF,

    // Identifiers
    IDENTIFIER,
    INTEGER,

    // Operators
    ASSIGN,
    PLUS,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    OPENING_SQUARE_BRACKET,
    CLOSING_SQUARE_BRACKET,

    OPENING_ANGLE_BRACKET,
    CLOSING_ANGLE_BRACKET,

    //Keywords
    FUNCTION,
    LET,
}
