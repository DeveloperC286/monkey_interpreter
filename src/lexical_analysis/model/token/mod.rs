#[derive(Debug, Clone, PartialEq)]
pub(crate) enum Token {
    // Identifiers
    Identifier { literal: String },

    //Values
    Integer { literal: i64 },
    True,
    False,
    String { literal: String },

    // Operators
    Assign,
    Plus,
    Minus,
    Divide,
    Multiply,
    GreaterThan,
    LesserThan,
    Not,
    Equals,
    NotEquals,

    // Delimiters
    Comma,
    SemiColon,

    // Brackets
    OpeningRoundBracket,
    ClosingRoundBracket,

    OpeningCurlyBracket,
    ClosingCurlyBracket,

    OpeningSquareBracket,
    ClosingSquareBracket,

    //Keywords
    Function,
    Let,
    If,
    Else,
    Return,
}
