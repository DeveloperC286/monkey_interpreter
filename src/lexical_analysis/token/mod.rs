use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[allow(non_camel_case_types)]
pub enum Token {
    ILLEGAL { literal: String },
    EOF,

    // Identifiers
    IDENTIFIER { literal: String },
    INTEGER { literal: String },

    // Operators
    ASSIGN,
    PLUS,
    MINUS,
    DIVIDE,
    MULTIPLY,
    GREATER_THAN,
    LESSER_THAN,
    NOT,
    EQUALS,
    NOT_EQUALS,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    //Keywords
    FUNCTION,
    LET,
    TRUE,
    FALSE,
    IF,
    ELSE,
    RETURN,
}
