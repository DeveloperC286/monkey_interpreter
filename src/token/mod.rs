#[derive(Debug, PartialEq)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

#[derive(Debug, Copy, Clone, PartialEq)]
#[allow(non_camel_case_types)]
pub enum TokenType {
    ILLEGAL,
    EOF,

    // Identifiers
    IDENTIFIER,
    INTEGER,

    // Operators
    ASSIGN,
    PLUS,
    MINUS,
    DIVIDE,
    MULTIPLY,
    GREATER_THAN,
    LESSER_THAN,

    // Delimiters
    COMMA,
    SEMI_COLON,

    // Brackets
    OPENING_ROUND_BRACKET,
    CLOSING_ROUND_BRACKET,

    OPENING_CURLY_BRACKET,
    CLOSING_CURLY_BRACKET,

    //Keywords
    FUNCTION,
    LET,
}
