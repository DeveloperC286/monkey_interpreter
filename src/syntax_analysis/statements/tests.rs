use insta::assert_json_snapshot;
use rstest::rstest;

use crate::lexical_analysis::token::Token;

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec ! [
            Token::RETURN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_return_statements_case1"
        ),
    case(
        vec ! [
            Token::RETURN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::RETURN,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_return_statements_case2"
    ),
)]
fn test_syntax_analysis_for_return_statements(tokens: Vec<Token>, snapshot_name: &str) {
    assert_expected_returned_abstract_syntax_tree!(tokens, snapshot_name);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec ! [
            Token::LET,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_let_statements_case1"
    ),
    case(
        vec ! [
            Token::LET,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::LET,
            Token::IDENTIFIER{literal: "z".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "7".to_string()},
            Token::PLUS,
            Token::INTEGER{literal: "10".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_let_statements_case2"
    ),
)]
fn test_syntax_analysis_for_let_statements(tokens: Vec<Token>, snapshot_name: &str) {
    assert_expected_returned_abstract_syntax_tree!(tokens, snapshot_name);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec ! [
            Token::LET,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::EOF,
        ],
        "test_syntax_analysis_for_let_statements_syntax_errors_case1"
    ),
    case(
        vec ! [
            Token::LET,
            Token::EOF,
        ],
        "test_syntax_analysis_for_let_statements_syntax_errors_case2"
    ),
)]
fn test_syntax_analysis_for_let_statements_syntax_errors(tokens: Vec<Token>, snapshot_name: &str) {
    assert_expected_returned_abstract_syntax_tree!(tokens, snapshot_name);
}

#[rstest(
    tokens_1,
    snapshot_name_1,
    tokens_2,
    snapshot_name_2,
    case(
        vec ! [
            Token::LET,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_let_statements_case1",
        vec ! [
            Token::LET,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::LET,
            Token::IDENTIFIER{literal: "z".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "7".to_string()},
            Token::PLUS,
            Token::INTEGER{literal: "10".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_let_statements_case2"
    ),
)]
fn test_syntax_analysis_for_successive_parsing(
    tokens_1: Vec<Token>,
    snapshot_name_1: &str,
    tokens_2: Vec<Token>,
    snapshot_name_2: &str,
) {
    assert_expected_returned_abstract_syntax_tree!(tokens_1, snapshot_name_1);
    assert_expected_returned_abstract_syntax_tree!(tokens_2, snapshot_name_2);
}

#[rstest(
    tokens,
    snapshot_name,
    case(
        vec ! [
            Token::LET,
            Token::IDENTIFIER{literal: "x".to_string()},
            Token::ASSIGN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::RETURN,
            Token::INTEGER{literal: "5".to_string()},
            Token::SEMI_COLON,
            Token::EOF,
        ],
        "test_syntax_analysis_for_combined_statements_case1"
    ),
)]
fn test_syntax_analysis_for_combined_statements(tokens: Vec<Token>, snapshot_name: &str) {
    assert_expected_returned_abstract_syntax_tree!(tokens, snapshot_name);
}
