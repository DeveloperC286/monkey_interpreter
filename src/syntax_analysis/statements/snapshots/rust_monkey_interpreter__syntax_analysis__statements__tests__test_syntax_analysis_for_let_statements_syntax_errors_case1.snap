---
source: src/syntax_analysis/statements/tests.rs
expression: returned_abstract_syntax_tree
---
{
  "abstract_syntax_tree": [],
  "syntax_parsing_errors": [
    "Syntax error : Expected a ASSIGN."
  ]
}
